plugins/soundsourcem4a/soundsourcem4a.cpp

```diff
@@ -1,221 +1,461 @@
+        }
     }
+    return MP4_INVALID_TRACK_ID;
 }
 
-Result SoundSourceM4A::open()
-{
-    //Initialize the FAAD2 decoder...
-    initializeDecoder();
+} // anonymous namespace
 
-    //qDebug() << "SSM4A: channels:" << getChannels()
-    //         << "filelength:" << filelength
-    //         << "Sample Rate:" << m_iSampleRate;
-    return OK;
+QList<QString> SoundSourceM4A::supportedFileExtensions() {
+    QList<QString> list;
+    list.push_back("m4a");
+    list.push_back("mp4");
+    return list;
+}
+
+SoundSourceM4A::SoundSourceM4A(QUrl url)
+        : SoundSourcePlugin(url, "m4a"),
+          m_hFile(MP4_INVALID_FILE_HANDLE),
+          m_trackId(MP4_INVALID_TRACK_ID),
+          m_maxSampleBlockId(MP4_INVALID_SAMPLE_ID),
+          m_curSampleBlockId(MP4_INVALID_SAMPLE_ID),
+          m_inputBufferOffset(0),
+          m_inputBufferLength(0),
+          m_hDecoder(NULL),
+          m_sampleBufferReadOffset(0),
+          m_sampleBufferWriteOffset(0),
+          m_curFrameIndex(kFrameIndexMin) {
 }
 
-int SoundSourceM4A::initializeDecoder()
-{
-    // Copy QString to char[] buffer for mp4_open to read from later
-    const QByteArray qbaFileName(getFilename().toLocal8Bit());
-    int bytes = qbaFileName.length() + 1;
-    ipd.filename = new char[bytes];
-    strncpy(ipd.filename, qbaFileName.constData(), bytes);
-    ipd.filename[bytes-1] = '\0';
-    ipd.remote = false; // File is not an stream
-    // The file was loading and failing erratically because
-    // ipd.remote was an in an uninitialized state, it needed to be
-    // set to false.
-
-    int mp4_open_status = mp4_open(&ipd);
-    if (mp4_open_status != 0) {
-        qWarning() << "SSM4A::initializeDecoder failed"
-                 << getFilename() << " with status:" << mp4_open_status;
+SoundSourceM4A::~SoundSourceM4A() {
+    close();
+}
+
+Result SoundSourceM4A::open() {
+    if (MP4_INVALID_FILE_HANDLE != m_hFile) {
+        qWarning() << "Cannot reopen M4A file:" << getUrl();
         return ERR;
     }
 
-    // mp4_open succeeded -> populate variables
-    mp4_private* mp = (struct mp4_private*)ipd.private_ipd;
-    DEBUG_ASSERT_AND_HANDLE(mp) {
-        mp4_close(&ipd);
+    /* open MP4 file, check for >= ver 1.9.1 */
+#if MP4V2_PROJECT_version_hex <= 0x00010901
+    m_hFile = MP4Read(getLocalFileNameBytes().constData(), 0);
+#else
+    m_hFile = MP4Read(getLocalFileNameBytes().constData());
+#endif
+    if (MP4_INVALID_FILE_HANDLE == m_hFile) {
+        qWarning() << "Failed to open file for reading:" << getUrl();
         return ERR;
     }
 
-    mp4file = mp->mp4.handle;
-    filelength = mp4_total_samples(&ipd);
-    setSampleRate(mp->sample_rate);
-    setChannels(mp->channels);
+    m_trackId = findFirstAudioTrackId(m_hFile);
+    if (MP4_INVALID_TRACK_ID == m_trackId) {
+        qWarning() << "No AAC track found:" << getUrl();
+        return ERR;
+    }
 
-    return OK;
-}
+    m_maxSampleBlockId = MP4GetTrackNumberOfSamples(m_hFile, m_trackId);
```

Comment:
This implicites a kSampleBlockIdMin = 1; 
I am slightly just removing the conditional code around kSampleBlockIdMin and just silently assume a 1 base. Otherwithe kSampleBlockIdMin should be used consequent.